packages/sso/src/oidc/discovery.ts

```diff
@@ -0,0 +1,390 @@
+ *
+ * Phase 1: Only validates that the URL is well-formed.
+ * Future phases will add SSRF protection, allowlist checks, etc.
+ *
+ * @param url - The discovery URL to validate
+ * @throws DiscoveryError if URL is invalid
+ */
+export function validateDiscoveryUrl(url: string): void {
+	try {
+		const parsed = new URL(url);
+		if (parsed.protocol !== "https:" && parsed.protocol !== "http:") {
+			throw new DiscoveryError(
+				"discovery_invalid_url",
+				`Discovery URL must use HTTP or HTTPS protocol: ${url}`,
+				{ url, protocol: parsed.protocol },
+			);
+		}
+	} catch (error) {
+		if (error instanceof DiscoveryError) {
+			throw error;
+		}
+		throw new DiscoveryError(
+			"discovery_invalid_url",
+			`Invalid discovery URL: ${url}`,
+			{
+				url,
+				originalError: error instanceof Error ? error.message : String(error),
+			},
+		);
+	}
+}
+
+/**
+ * Fetch the OIDC discovery document from the IdP.
+ *
+ * @param url - The discovery endpoint URL
+ * @param timeout - Request timeout in milliseconds
+ * @returns The parsed discovery document
+ * @throws DiscoveryError on network errors, timeouts, or invalid responses
+ */
+export async function fetchDiscoveryDocument(
+	url: string,
+	timeout: number = DEFAULT_DISCOVERY_TIMEOUT,
+): Promise<OIDCDiscoveryDocument> {
+	try {
+		const controller = new AbortController();
+		const timeoutId = setTimeout(() => controller.abort(), timeout);
+
+		const response = await betterFetch<OIDCDiscoveryDocument>(url, {
+			method: "GET",
+			signal: controller.signal,
+		});
+
+		clearTimeout(timeoutId);
+
+		if (response.error) {
+			const status = response.error.status;
+
+			if (status === 404) {
+				throw new DiscoveryError(
+					"discovery_not_found",
+					`Discovery endpoint not found: ${url}`,
+					{ status },
+				);
+			}
+
+			throw new DiscoveryError(
+				"discovery_unexpected_error",
+				`Discovery request failed with status ${status}: ${response.error.message}`,
+				{ status, message: response.error.message },
+			);
+		}
+
+		if (!response.data) {
+			throw new DiscoveryError(
+				"discovery_invalid_json",
+				"Discovery endpoint returned empty response",
+				{ url },
+			);
+		}
+
+		return response.data;
+	} catch (error) {
+		if (error instanceof DiscoveryError) {
+			throw error;
+		}
+
+		if (error instanceof Error && error.name === "AbortError") {
+			throw new DiscoveryError(
+				"discovery_timeout",
+				`Discovery request timed out after ${timeout}ms`,
+				{ url, timeout },
+			);
+		}
+
+		if (error instanceof SyntaxError) {
+			throw new DiscoveryError(
+				"discovery_invalid_json",
+				"Discovery endpoint returned invalid JSON",
+				{ url, originalError: error.message },
+			);
+		}
+
+		throw new DiscoveryError(
+			"discovery_unexpected_error",
+			`Unexpected error during discovery: ${error instanceof Error ? error.message : String(error)}`,
+			{
+				url,
+				originalError: error instanceof Error ? error.message : String(error),
```

Comment:
I wonder if we want to use error-cause instead to expose the whole causal chain instead of only the top-most error?